src/openssl/ssl/context.cr

```diff
@@ -109,6 +109,90 @@ abstract class OpenSSL::SSL::Context
   end
 
   class Server < Context
+    protected def set_tlsext_servername_callback
+      cb = ->(ssl : LibSSL::SSL, cmd : LibC::Int, ctxptr : Void*) {
+        ctx = Box(Server).unbox(ctxptr)
+        if ctx.sni_fail_hard
+          ret = LibSSL::SSL_TLSEXT_ERR_OK
+        else
+          ret = LibSSL::SSL_TLSEXT_ERR_NOACK
+        end
+        hostname_ptr = LibSSL.ssl_get_servername(ssl, LibSSL::TLSExt::NAMETYPE_host_name)
+        hostname = if hostname_ptr
+                     String.new(hostname_ptr)
+                   else
+                     nil
+                   end
+        if hostname
+          if sniCtx = ctx.sni[hostname]?
+            LibSSL.ssl_set_ssl_ctx(ssl, sniCtx)
+            ret = LibSSL::SSL_TLSEXT_ERR_OK
+          else
+            ctx.sni.each do |sni, sniCtx|
+              if HostnameValidation.matches_hostname?(sni, hostname)
+                LibSSL.ssl_set_ssl_ctx(ssl, sniCtx)
+                ret = LibSSL::SSL_TLSEXT_ERR_OK
+                break
+              end
+            end
```

Comment:
I'm not sure these two branches should be the same interface, that is for an interface that asks me for hostnames explicitly I would expect the certificate to only be presented for that exact hostname.

On the other hand being able to just give a bunch of certificates and have the valid hostnames figured out from that is quite useful, so a separate interface for that does make sense to me. That should then clearly document that in case of certificates with overlapping hostnames, say you add one for example.org and example.com and one for example.org and example.net, the first one that matches win.

Finally I think mixing both approaches should also work but still a certificate that was added via the "figure the hostnames from it" interface shouldn't be presented for an explicitly given hostname, likewise an explicitly given hostname's certificate should always win over the implicit one if present.

Now this starts to accumulate quite some logic to the point where I start to wonder whether a `Context::Server` subclass makes sense, perhaps with some nice self yielding constructor so you can do something like

```cr
server.context = OpenSSL::SSL::Context::SniServer.new do |ctx|
  ctx.add_certifcate cert_a, key_a
  ctx.add_certificate "example.org", cert_b, key_b
  ctx.add_certificate "example.net", some_context
  ctx.add_certificate another_context # not sure about that one, would need to validate the context has a certificate etc.
end
```

I don't want to ask you to do all of that and I am mostly brainstorming here. But if there's some general agreement along those thoughts we should maybe figure out what the minimal thing is we can do here without introducing weird behaviours that we have to break later.